packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -410,106 +570,655 @@ class _CupertinoAlertActionSection extends StatelessWidget {
+  // + 50% the height of the 2nd button + 2 dividers.
+  double _computeMinIntrinsicHeightStacked(double width) {
+    assert(childCount >= 3);
+
+    final List<RenderBox> children = getChildrenAsList();
+    return (2 * dividerThickness)
+        + children[0].computeMinIntrinsicHeight(width)
+        + (0.5 * children[1].computeMinIntrinsicHeight(width));
+  }
+
+  @override
+  double computeMaxIntrinsicHeight(double width) {
+    if (childCount == 0) {
+      // No buttons. Zero height.
+      return 0.0;
+    } else if (childCount == 1) {
+      // One button. Our max intrinsic height is equal to the button's.
+      return firstChild.computeMaxIntrinsicHeight(width) + dividerThickness;
+    } else if (childCount == 2) {
+      // Two buttons...
+      if (_isSingleButtonRow(width)) {
+        // The 2 buttons fit side by side so our max intrinsic height is equal
+        // to the taller of the 2 buttons.
+        final double perButtonWidth = (width - dividerThickness) / 2.0;
+        return math.max(
+          firstChild.computeMaxIntrinsicHeight(perButtonWidth),
+          lastChild.computeMaxIntrinsicHeight(perButtonWidth),
+        ) + dividerThickness;
+      } else {
+        // The 2 buttons do not fit side by side. Measure total height as a
+        // vertical stack.
+        return _computeMaxIntrinsicHeightStacked(width);
+      }
+    } else {
+      // Three+ buttons. Stack the buttons vertically with dividers and measure
+      // the overall height.
+      return _computeMaxIntrinsicHeightStacked(width);
+    }
+  }
+
+  // Max height of a stack of buttons is the sum of all button heights + a
+  // divider for each button.
+  double _computeMaxIntrinsicHeightStacked(double width) {
+    assert(childCount >= 2);
+
+    final double allDividersHeight = childCount * dividerThickness;
+    return getChildrenAsList().fold(allDividersHeight, (double heightAccum, RenderBox button) {
+      return heightAccum + button.computeMaxIntrinsicHeight(width);
+    });
+  }
+
+  bool _isSingleButtonRow(double width) {
+    if (childCount == 1) {
+      return true;
+    } else if (childCount == 2) {
+      // There are 2 buttons. If they can fit side-by-side then that's what
+      // we want to do. Otherwise, stack them vertically.
+      final double sideBySideWidth = firstChild.computeMaxIntrinsicWidth(double.infinity)
+          + dividerThickness
+          + lastChild.computeMaxIntrinsicWidth(double.infinity);
+      return sideBySideWidth <= width;
+    } else {
+      return false;
+    }
+  }
+
+  @override
+  void performLayout() {
+    if (_isSingleButtonRow(_kCupertinoDialogWidth)) {
+      if (childCount == 1) {
+        // We have 1 button. Our size is the width of the dialog and the height
+        // of the single button.
+        firstChild.layout(
+          constraints,
+          parentUsesSize: true,
+        );
+
+        size = new Size(_kCupertinoDialogWidth, firstChild.size.height + dividerThickness);
```

Comment:
here and elsewhere: when setting the height, you don't seem to guarantee that the constraints are met.

this needs a test

Reply:
I applied incoming constraints to the sizes of both the overall dialog render object, as well as the actions render object. I added a test that constrains the overall dialog to a very small 200.0 by 100.0 which is too small in both dimensions.